File: klinara-ios/klinara-ios/Features/Packages/PackageDefinitionForm.swift
import SwiftUI

/// ``PackageDefinitionEditorView``'in düzenleme durumu.
///
/// ``ServiceForm`` kalıbı: gözlemlenebilir olmayan bir değer tipi, `@State`
/// içinde yaşar, `isValid`/`isDirty`'yi kendi hesaplar ve iki ayrı wire
/// gövdesini (`Create`/`Update`) kendi kurar.
struct PackageDefinitionForm {

    /// Kalemin form içindeki hâli. Sunucu gövdesi yalnız `serviceId` + `quantity`
    /// taşır ama ekranın ada ve liste fiyatına da ihtiyacı var — indirim
    /// önizlemesi bunlar olmadan hesaplanamaz.
    struct Item: Identifiable, Equatable {
        let id = UUID()
        var serviceId: String
        var serviceName: String
        var unitListPriceMinor: Int
        var quantity: Int

        var listTotalMinor: Int { unitListPriceMinor * quantity }
    }

    var slug: String
    var name: String
    var description: String
    var totalPriceMinor: Int?
    var branchId: String?
    /// `nil` **süresiz** paket demektir; `0` değil.
    var validityDays: Int?
    var isTransferable: Bool
    var isOnlineSellable: Bool
    var isActive: Bool
    var items: [Item]

    private var slugIsCustom: Bool
    private let original: Snapshot
    /// Düzenlemede slug ve şube kilitlidir: ikisi de satılmış paketlerin
    /// izini etkiler ve sunucu `PATCH` gövdesinde ikisini de kabul etmez.
    let isEditing: Bool

    private struct Snapshot: Equatable {
        var slug: String
        var name: String
        var description: String
        var totalPriceMinor: Int?
        var branchId: String?
        var validityDays: Int?
        var isTransferable: Bool
        var isOnlineSellable: Bool
        var isActive: Bool
        var items: [Item]
    }

    init(existing: PackageDefinition?) {
        slug = existing?.slug ?? ""
        name = existing?.name ?? ""
        description = existing?.description ?? ""
        totalPriceMinor = existing?.totalPriceMinor
        branchId = existing?.branchId
        validityDays = existing?.validityDays
        isTransferable = existing?.isTransferable ?? true
        isOnlineSellable = existing?.isOnlineSellable ?? false
        isActive = existing?.isActive ?? true
        items = (existing?.items ?? []).sorted { $0.sortOrder < $1.sortOrder }.map {
            Item(
                serviceId: $0.serviceId,
                serviceName: $0.serviceName,
                unitListPriceMinor: $0.unitListPriceMinor,
                quantity: $0.quantity
            )
        }
        slugIsCustom = existing != nil
        isEditing = existing != nil

        original = Snapshot(
            slug: slug, name: name, description: description,
            totalPriceMinor: totalPriceMinor, branchId: branchId,
            validityDays: validityDays, isTransferable: isTransferable,
            isOnlineSellable: isOnlineSellable, isActive: isActive, items: items
        )
    }

    // MARK: Türetilmiş

    private var current: Snapshot {
        Snapshot(
            slug: slug, name: name, description: description,
            totalPriceMinor: totalPriceMinor, branchId: branchId,
            validityDays: validityDays, isTransferable: isTransferable,
            isOnlineSellable: isOnlineSellable, isActive: isActive, items: items
        )
    }

    var isDirty: Bool { current != original }

    var isValid: Bool {
        !name.trimmingCharacters(in: .whitespaces).isEmpty
            && Slug.isValid(slug)
            && totalPriceMinor != nil
            && !items.isEmpty
            && items.allSatisfy { $0.quantity > 0 }
            && !hasDuplicateService
    }

    var slugValidationMessage: String? {
        guard !slug.isEmpty, !Slug.isValid(slug) else { return nil }
        return "Yalnız küçük harf, rakam ve tire; 3-50 karakter."
    }

    /// Sunucu aynı hizmetin iki kez verilmesini reddediyor; kullanıcıya `400`
    /// yerine kırmızı bir satır göstermek daha dürüst.
    var hasDuplicateService: Bool {
        Set(items.map(\.serviceId)).count != items.count
    }

    var totalSessions: Int { items.reduce(0) { $0 + $1.quantity } }

    /// Kalemlerin **güncel** katalog fiyatları toplamı. Satış fiyatıyla farkı
    /// kullanıcıya indirimi gösterir; sunucudaki `listPriceMinor` ile aynı hesap.
    var listPriceMinor: Int { items.reduce(0) { $0 + $1.listTotalMinor } }

    var discountMinor: Int? {
        guard let totalPriceMinor else { return nil }
        let diff = listPriceMinor - totalPriceMinor
        return diff > 0 ? diff : nil
    }

    // MARK: Düzenleme

    mutating func nameDidChange(_ newValue: String) {
        guard !slugIsCustom else { return }
        slug = Slug.make(from: newValue)
    }

    mutating func slugDidChange(_ newValue: String) {
        guard !slugIsCustom, newValue != Slug.make(from: name) else { return }
        slugIsCustom = true
    }

    mutating func add(service: ClinicService, in branchId: String?) {
        guard !items.contains(where: { $0.serviceId == service.id }) else { return }
        let effective = service.effective(in: branchId)
        items.append(
            Item(
                serviceId: service.id,
                serviceName: service.name,
                unitListPriceMinor: effective.priceMinor,
                quantity: 1
            )
        )
    }

    mutating func remove(_ item: Item) {
        items.removeAll { $0.id == item.id }
    }

    // MARK: Sunucu gövdeleri

    private var wireItems: [PackageDefinitionItemInput] {
        items.map { PackageDefinitionItemInput(serviceId: $0.serviceId, quantity: $0.quantity) }
    }

    private var trimmedDescription: String {
        description.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func createInput() -> CreatePackageDefinitionInput {
        CreatePackageDefinitionInput(
            slug: slug,
            name: name.trimmingCharacters(in: .whitespaces),
            description: trimmedDescription.isEmpty ? nil : trimmedDescription,
            totalPriceMinor: totalPriceMinor ?? 0,
            branchId: branchId,
            validityDays: validityDays,
            isTransferable: isTransferable,
            isOnlineSellable: isOnlineSellable,
            isActive: isActive,
            items: wireItems
        )
    }

    /// Kalem listesi **her zaman** gönderilir: sunucu verilen listeyle
    /// tamamen değiştiriyor ve formda kalemler zaten tam hâlleriyle duruyor.
    func updateInput() -> UpdatePackageDefinitionInput {
        UpdatePackageDefinitionInput(
            name: name.trimmingCharacters(in: .whitespaces),
            description: .text(trimmedDescription),
            totalPriceMinor: totalPriceMinor,
            // "Süresiz" seçildiyse alan TEMİZLENİR; gönderilmemesi eski süreyi
            // olduğu gibi bırakırdı.
            validityDays: validityDays.map { Nullable.set($0) } ?? .clear,
            isTransferable: isTransferable,
            isOnlineSellable: isOnlineSellable,
            isActive: isActive,
            items: wireItems
        )
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/features/packages/PackageDefinitionForm.kt
package com.klinara.android.features.packages

import com.klinara.android.services.catalog.ClinicService
import com.klinara.android.services.crm.Patch
import com.klinara.android.services.formatting.Slug
import com.klinara.android.services.packages.CreatePackageDefinitionInput
import com.klinara.android.services.packages.PackageDefinition
import com.klinara.android.services.packages.PackageDefinitionItemInput
import com.klinara.android.services.packages.UpdatePackageDefinitionInput

/**
 * Paket tanımı formunun durumu — **saf bir değer tipi** (`CustomerForm` deseni).
 *
 * Kurallar ekranda değil burada: geçerlilik, kirlilik, indirim önizlemesi ve iki ayrı
 * gövdenin (`Create`/`Update`) üretimi `@Composable` olmayan bir tipte durduğu için testi
 * ucuz ve oluşturma/düzenleme aynı yoldan sürülüyor. iOS `PackageDefinitionForm` paritesi.
 */
data class PackageDefinitionForm(
    val slug: String = "",
    val name: String = "",
    val description: String = "",
    val totalPriceMinor: Long? = null,
    /** `null` **tüm şubeler**. */
    val branchId: String? = null,
    /** `null` **süresiz** paket demektir; `0` değil. */
    val validityDays: Int? = null,
    val isTransferable: Boolean = true,
    val isOnlineSellable: Boolean = false,
    val isActive: Boolean = true,
    val items: List<Item> = emptyList(),
    /**
     * Düzenlemede slug ve şube KİLİTLİ: ikisi de satılmış paketlerin izini taşır ve sunucu
     * `PATCH` gövdesinde ikisini de reddeder.
     */
    val isEditing: Boolean = false,
    /** Kullanıcı slug'a elle dokunduysa ad değişince artık türetilmez. */
    private val slugIsCustom: Boolean = false,
    private val original: PackageDefinitionForm? = null,
) {
    /**
     * Kalemin form içindeki hâli. Sunucu gövdesi yalnız `serviceId` + `quantity` taşır ama
     * ekranın ada ve liste fiyatına da ihtiyacı var — indirim önizlemesi bunlarsız
     * hesaplanamaz.
     */
    data class Item(
        val serviceId: String,
        val serviceName: String,
        val unitListPriceMinor: Long,
        val quantity: Int,
    ) {
        val listTotalMinor: Long get() = unitListPriceMinor * quantity
    }

    // --- Türetilmiş ---

    val isDirty: Boolean get() = original == null || comparable() != original.comparable()

    val slugError: String?
        get() = if (slug.isEmpty() || SLUG_PATTERN.matches(slug)) null else SLUG_HINT

    val validityError: String?
        get() = validityDays?.takeIf { it < 1 }?.let { "En az 1 gün; süresiz için anahtarı kapatın." }

    /** Sunucu aynı hizmetin iki kez verilmesini reddediyor; ekranda söylemek 400'den dürüst. */
    val hasDuplicateService: Boolean get() = items.map { it.serviceId }.toSet().size != items.size

    val isValid: Boolean
        get() =
            name.isNotBlank() && SLUG_PATTERN.matches(slug) && totalPriceMinor != null &&
                items.isNotEmpty() && items.all { it.quantity > 0 } && !hasDuplicateService &&
                validityError == null

    val totalSessions: Int get() = items.sumOf { it.quantity }

    /** Kalemlerin GÜNCEL katalog fiyatları toplamı — sunucudaki `listPriceMinor` ile aynı hesap. */
    val listPriceMinor: Long get() = items.sumOf { it.listTotalMinor }

    /** `null` indirim yok demek; sıfır göstermek "indirim var ama 0 ₺" gibi okunurdu. */
    val discountMinor: Long?
        get() = totalPriceMinor?.let { price -> (listPriceMinor - price).takeIf { it > 0 } }

    // --- Düzenleme ---

    fun withName(value: String): PackageDefinitionForm =
        copy(name = value, slug = if (slugIsCustom || isEditing) slug else Slug.make(value))

    fun withSlug(value: String): PackageDefinitionForm =
        copy(slug = value, slugIsCustom = value != Slug.make(name))

    fun adding(
        service: ClinicService,
        branchId: String?,
    ): PackageDefinitionForm {
        if (items.any { it.serviceId == service.id }) return this
        val item =
            Item(
                serviceId = service.id,
                serviceName = service.name,
                // Şubenin geçerli fiyatı — indirim önizlemesi seçili şubeye göre doğru olsun.
                unitListPriceMinor = service.effective(branchId).priceMinor,
                quantity = 1,
            )
        return copy(items = items + item)
    }

    fun removing(serviceId: String): PackageDefinitionForm = copy(items = items.filterNot { it.serviceId == serviceId })

    fun withQuantity(
        serviceId: String,
        quantity: Int,
    ): PackageDefinitionForm =
        copy(items = items.map { if (it.serviceId == serviceId) it.copy(quantity = quantity) else it })

    // --- Sunucu gövdeleri ---

    private val wireItems: List<PackageDefinitionItemInput>
        get() = items.map { PackageDefinitionItemInput(it.serviceId, it.quantity) }

    fun createInput(): CreatePackageDefinitionInput =
        CreatePackageDefinitionInput(
            slug = slug,
            name = name.trim(),
            totalPriceMinor = totalPriceMinor ?: 0,
            items = wireItems,
            description = description.trim().ifEmpty { null },
            branchId = branchId,
            validityDays = validityDays,
            isTransferable = isTransferable,
            isOnlineSellable = isOnlineSellable,
            isActive = isActive,
        )

    /**
     * Güncelleme gövdesi — yalnız DEĞİŞEN alanlar.
     *
     * Kalem listesi değiştiyse **tamamı** gönderilir: sunucu verilen listeyle tamamen
     * değiştiriyor. "Süresiz" seçildiyse `validityDays` açıkça TEMİZLENİR; gönderilmemesi
     * eski süreyi olduğu gibi bırakırdı.
     */
    fun updateInput(): UpdatePackageDefinitionInput {
        val base = original ?: return UpdatePackageDefinitionInput()
        return UpdatePackageDefinitionInput(
            name = name.trim().takeIf { it != base.name },
            description = if (description.trim() != base.description) Patch.text(description) else Patch.Unchanged,
            totalPriceMinor = totalPriceMinor.takeIf { it != base.totalPriceMinor },
            validityDays = if (validityDays != base.validityDays) Patch.orClear(validityDays) else Patch.Unchanged,
            isTransferable = isTransferable.takeIf { it != base.isTransferable },
            isOnlineSellable = isOnlineSellable.takeIf { it != base.isOnlineSellable },
            isActive = isActive.takeIf { it != base.isActive },
            items = wireItems.takeIf { items != base.items },
        )
    }

    /** Kirlilik karşılaştırması `original`/`slugIsCustom`'u dışarıda bırakır. */
    private fun comparable() = copy(original = null, slugIsCustom = false)

    companion object {
        /** Sunucunun `SLUG_PATTERN`'i. */
        private val SLUG_PATTERN = Regex("^[a-z0-9][a-z0-9-]{1,48}[a-z0-9]$")
        private const val SLUG_HINT = "Yalnız küçük harf, rakam ve tire; 3-50 karakter."

        fun empty(branchId: String? = null): PackageDefinitionForm = PackageDefinitionForm(branchId = branchId)

        fun of(definition: PackageDefinition): PackageDefinitionForm {
            val form =
                PackageDefinitionForm(
                    slug = definition.slug,
                    name = definition.name,
                    description = definition.description.orEmpty(),
                    totalPriceMinor = definition.totalPriceMinor,
                    branchId = definition.branchId,
                    validityDays = definition.validityDays,
                    isTransferable = definition.isTransferable,
                    isOnlineSellable = definition.isOnlineSellable,
                    isActive = definition.isActive,
                    items =
                        definition.sortedItems.map {
                            Item(it.serviceId, it.serviceName, it.unitListPriceMinor, it.quantity)
                        },
                    isEditing = true,
                    slugIsCustom = true,
                )
            return form.copy(original = form)
        }
    }
}
